book/accessibility.md

```diff
@@ -1812,65 +1801,77 @@ class Browser:
-the browser thread doesn't actually know what in the `Tab` is focused, so we'll
-need to commit that to a `tab_focus` field on `Browser` as well.^[By adding a
-`focus` field to `CommitData`, etc., just like the accessibilty tree, and
-storing it in a new `tab_focus` property on the `Browser`. Since it's
-repetitive I've omitted that code.]
+Speaking the whole document happens only once. But the user might need
+feedback as they browse the page. For example, let's speak the focused
+element to the user.
+
+To do that, the browser thread is going to need to know which element
+is focused. Let's add that to the `CommitData`; I'm not going to show
+the code, because it's repetitive, but the point is to store the
+`Tab`'s `focus` field in the `Browser`'s `tab_focus` field.
 
-Over on the browser, a new `speak_node` method will do something similar to
-`speak_document`, but specific to one node:
+Now we need to know when focus changes. While we could do that lots of
+ways,[^eg-handle-tab] the best way is to store a `last_tab_focus`
+field on `Browser` with the last focused element we actually spoke out
+loud:
 
 ``` {.python}
 class Browser:
-    def speak_node(self, node, text):
-        text += node.text
-        if text and node.children and \
-            node.children[0].role == "StaticText":
-            text += " " + \
-            node.children[0].text
-
-        print(text)
-        if text:
-            if not self.is_muted():
-                speak_text(text)
+    def __init__(self):
+        # ...
+        self.last_tab_focus = None
 ```
 
-
-And then we can simply call it from `speak_update`:
+Then, if `tab_focus` isn't equal to `last_tab_focus`, we know focus
+has moved and it's time to speak the focused node:
 
 ``` {.python}
 class Browser:
-    def __init__(self):
-        # ...
-        self.accessibility_focus = None
-
     def speak_update(self):
         if self.tab_focus and \
-            self.tab_focus != self.accessibility_focus:
+            self.tab_focus != self.last_tab_focus:
             nodes = [node for node in tree_to_list(self.accessibility_tree, [])
                         if node.node == self.tab_focus]
             if nodes:
-                self.accessibility_focus = self.tab_focus
-            self.speak_node(nodes[0], "element focused ")
+                self.speak_node(nodes[0], "element focused ")
+            self.last_tab_focus = self.tab_focus
 ```
 
+The `speak_node` method is similar to `speak_document` but it only
+speaks a single node:
+
+``` {.python}
+class Browser:
+    def speak_node(self, node, text):
+        text += node.text
+        if text and node.children and \
+            node.children[0].role == "StaticText":
+            text += " " + \
+            node.children[0].text
+
+        if text:
+            speak_text(text)
+```
+
+There's a lot more in a real screen reader: landmarks, navigating text
+at different granularities, repeating recently spoken text, and so on.
```

Comment:
*not* repeating recently spoken text?

Reply:
Screen readers have key bindings for "repeat that word again" or "repeat that sentence you just said", which is what I meant to point to here. Can you suggest a better phrasing?